File: classic150/Solution0025.cpp
#include <deque>
using namespace std;
// 给你链表的头节点 head ，每 k 个节点一组进行翻转，请你返回修改后的链表。

// k 是一个正整数，它的值小于或等于链表的长度。如果节点总数不是 k 的整数倍，那么请将最后剩余的节点保持原有顺序。

// 你不能只是单纯的改变节点内部的值，而是需要实际进行节点交换。

class Solution
{
private:
    struct ListNode
    {
        int val;
        ListNode *next;
        ListNode(int x) : val(x), next(nullptr) {}
        ListNode(int x, ListNode *next) : val(x), next(next) {}
    };

public:
    ListNode *reverseKGroup(ListNode *head, int k)
    {
        if (!head)
            return nullptr;
        ListNode *prev = head, *next = head->next;
        int count = 0;
        while (++count < k && next != nullptr) // 翻转k-1次
        {
            prev->next = next->next;
            next->next = head;
            head = next;
            next = prev->next;
        }
        if (count < k)
        {
            return reverseKGroup(head, count); // 如果不足k个，翻转回去
        }
        prev->next = reverseKGroup(next, k);
        return head;
    }
};

class Solution
{
private:
    struct ListNode
    {
        int val;
        ListNode *next;
        ListNode(int x) : val(x), next(nullptr) {}
        ListNode(int x, ListNode *next) : val(x), next(next) {}
    };
    deque<ListNode *> deque;

public:
    ListNode *reverseKGroup(ListNode *head, int k)
    {
        ListNode *current = head;
        ListNode *newHead = nullptr, *prev = nullptr;
        int count = 0;
        while (current != nullptr)
        {
            deque.push_back(current);
            current = current->next;
            count++;
            if (count == k)
            {
                while (!deque.empty())
                {
                    ListNode *node = deque.back();
                    deque.pop_back();
                    if (prev == nullptr)
                    {
                        newHead = node;
                    }
                    else
                    {
                        prev->next = node;
                    }
                    prev = node;
                }
                prev->next = nullptr;
                count = 0;
            }
        }
        if (count < k)
        {
            while (!deque.empty())
            {
                ListNode *node = deque.front();
                deque.pop_front();
                if (prev == nullptr)
                {
                    newHead = node;
                }
                else
                {
                    prev->next = node;
                }
                prev = node;
            }
        }
        return newHead;
    }
};

class Solution
{
private:
    struct ListNode
    {
        int val;
        ListNode *next;
        ListNode(int x) : val(x), next(nullptr) {}
        ListNode(int x, ListNode *next) : val(x), next(next) {}
    };

public:
    ListNode *reverseKGroup(ListNode *head, int k)
    {
        ListNode *dummy = new ListNode(0, head);
        ListNode *prev = dummy;// prev是前一个k组的尾节点
        while (prev->next)
        {
            ListNode *node = reverse(prev, k);
            if (node == nullptr)
            {
                reverse(prev, k); // 不足k个，翻转回去
                break;
            }
            else
            {
                prev = node;
            }
        }
        return dummy->next;
    }

private:
    ListNode *reverse(ListNode *prev, int k)
    {
        ListNode *current = prev->next;
        int count = 1;
        for (; count < k && current != nullptr && current->next != nullptr; count++)
        {
            ListNode *next = current->next;
            current->next = next->next;
            next->next = prev->next;
            prev->next = next;
        }
        return (count == k) ? current : nullptr;
    }
};

File: anew/Solution0025.java
package anew;

import java.util.ArrayDeque;
import java.util.Deque;

public class Solution0025 {
    private static class ListNode {
        int val;
        ListNode next;

        public ListNode(int x) {
            val = x;
            next = null;
        }
    }

    public ListNode reverseKGroup(ListNode head, int k) {
        Deque<ListNode> deque = new ArrayDeque<>();
        ListNode dummy = new ListNode(0);
        ListNode current = dummy;
        ListNode node = head;
        while (node != null) {
            int count = 0;
            while (node != null && count < k) {
                deque.addLast(node);
                node = node.next;
                count++;
            }
            if (count == k) {
                while (!deque.isEmpty()) {
                    current.next = deque.pollLast();
                    current = current.next;
                }
                current.next = null;
            } else {
                while (!deque.isEmpty()) {
                    current.next = deque.pollFirst();
                    current = current.next;
                }
            }
        }
        return dummy.next;
    }

    public ListNode reverseKGroup2(ListNode head, int k) {
        if (head == null || k == 1)
            return head;
        // 三个指针：prev, curr, next
        ListNode prev = null;
        ListNode curr = head;
        int count = 0;
        while (curr != null && count < k) {
            ListNode next = curr.next;
            curr.next = prev;
            prev = curr; // prev 一直是当前翻转后的链表的头节点
            curr = next;
            count++;
        }
        if (count < k) {
            return reverseKGroup2(prev, count); // 如果不足 k 个节点，继续翻转回原来的顺序
        }
        head.next = reverseKGroup2(curr, k);
        return prev;
    }

    public ListNode reverseKGroup3(ListNode head, int k) {
        if (head == null || k == 1) {
            return head;
        }
        ListNode dummy = new ListNode(0);
        dummy.next = head;
        ListNode prevGroupEnd = dummy;
        while (true) {
            ListNode groupStart = prevGroupEnd.next;
            ListNode node = groupStart;
            for (int i = 1; i < k && node != null; i++) {
                node = node.next;
            }
            if (node == null) {
                break;
            }
            ListNode nextGropuStart = node.next;
            ListNode prev = null;
            ListNode curr = groupStart;
            for (int i = 0; i < k; i++) {
                ListNode next = curr.next;
                curr.next = prev;
                prev = curr;
                curr = next;
            }
            prevGroupEnd.next = prev;
            groupStart.next = nextGropuStart;
            prevGroupEnd = groupStart;
        }
        return dummy.next;
    }

    public static void main(String[] args) {
        Solution0025 solution = new Solution0025();
        ListNode head = new ListNode(1);
        head.next = new ListNode(2);
        head.next.next = new ListNode(3);
        head.next.next.next = new ListNode(4);
        head.next.next.next.next = new ListNode(5);

        int k = 3;
        ListNode newHead = solution.reverseKGroup3(head, k);
        System.out.print("翻转后的链表: ");
        while (newHead != null) {
            System.out.print(newHead.val + " ");
            newHead = newHead.next;
        }
    }
}
